### PR TITLE
Attachments sample c#

### PR DESCRIPTION
**Pull request by JasonSowers**

### CODE REVIEW

**Comment by cleemullins** [comment]
@JasonSowers The overall directory structure seems incorrect, as things are nested too deeply. Can you move the files up a level?

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Schema;
+using System.IO;
+using System.Net;
+using Microsoft.Bot.Connector;
+
+namespace Handling_Attachments
+{
+    /// <summary>
+    /// This bot will respond to the user's text input with an <see cref="Attachment"/> (in this case an image)
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs]: >will respond [](start = 17, length = 12)

response. Present tense is the best choice most of the time.

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+    /// using various types of attachments.  In this case, we are displaying an image from a file on the server,
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs]: >  [](start = 43, length = 1)

Please use only one space after a period. Apply this over the entire comment.

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    /// an image from an https url, and an uploaded image.  In this project the user also has the option to upload
+    /// an attachment to the bot.  The attachment will be saved to the the same directory as this file.
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs]: > will be save [](start = 49, length = 13)

(verb tense) is saved

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+    /// A message exchange between user and bot can contain media attachments, such as images, video, audio, and files.
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs]: >can [](start = 48, length = 3)

may

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+    /// The types of attachments that can be sent and recieved varies by channel.  In this sample we demonstrate sending
+    /// a <see cref="HeroCard"/> and images as attachments.  Also demonstrated is the ability of a bot to recieve file attachments.
+    /// </summary>
+    public class AtachmentsBot : IBot
+    {
+        /// <summary>
+        /// This controls what happens when an activity gets sent to the bot.
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs]: >activity [](start = 47, length = 8)

Activity. Add a see tag.

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+                    var reply = ProcessInput(turnContext);
+
+                    // Respond to the user.
+                    await turnContext.SendActivityAsync(reply, cancellationToken);
+
+                    await DisplayOptionsAsync(turnContext, cancellationToken);
+
+                    break;
+                case ActivityTypes.ConversationUpdate:
+                    // Send a welcome message to the user and tell them what actions they may perform to use this bot
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs]: >bot [](start = 114, length = 3)

Add a "." at the end.

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+                    if (turnContext.Activity.MembersAdded.Any())
+                    {
+                        await SendWelcomeMessageAsync(turnContext, cancellationToken);
+                    }
+
+                    break;
+            }
+        }
+
+        private static async Task DisplayOptionsAsync(ITurnContext turnContext, CancellationToken cancellationToke)
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs]: >cancellationToke [](start = 98, length = 16)

spelling. Toke != Token

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+        {
+            var reply = turnContext.Activity.CreateReply();
+
+            // Create a HeroCard with options for the user to choose to interact with the bot.
+            var card = new HeroCard
+            {
+                Text = "You can upload an image or select one of the following choices",
+                Buttons = new List<CardAction>()
+                {
+                    // Note that some channel require different values to be used in order to get buttons to display text.
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs]: >channel [](start = 38, length = 7)

plural. channels.

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+                    // In this code the emulator is accounted for with 'title', but in other channels you may need to provide a value
+                    // for other parameters like 'text' or 'displayText'.
+                    new CardAction(ActionTypes.ImBack, title: "1. Inline Attachment", value: "1"),
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs]: Strongly consider adding in a link here to the docs that have this breakdown.

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+            };
+
+            // Add the card to our reply.
+            reply.Attachments = new List<Attachment>() { card.ToAttachment() };
+
+            await turnContext.SendActivityAsync(reply, cancellationToke);
+        }
+
+        /// <summary>
+        /// On a conversation update activity sent to the bot, the bot will
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs]: >activity [](start = 37, length = 8)

Casing Activty. Add a See tag as well.

Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs]: >will [](start = 71, length = 4)

Verb tense. Make present tense. ("the bot sends"

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+        /// send a message to the any new user(s) that were added.
+        /// </summary>
+        /// <param name="turnContext">Provides the <see cref="ITurnContext"/> for the turn of the bot.</param>
+        /// <param name="cancellationToken" >(Optional) A <see cref="CancellationToken"/> that can be used by other objects
+        /// or threads to receive notice of cancellation.</param>
+        /// <returns>>A <see cref="Task"/> representing the operation result of the Turn operation.</returns>
+        private static async Task SendWelcomeMessageAsync(ITurnContext turnContext, CancellationToken cancellationToken)
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs]: >SendWelcomeMessageAsync [](start = 34, length = 23)

This code is different than the greeting code we have in other bots. What's missing here?

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+                        $" This bot will introduce you to Attachments." +
+                        $"  Please select an option",
+                        cancellationToken: cancellationToken);
+                    await DisplayOptionsAsync(turnContext, cancellationToken);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Given the input from the message activity the user sent, create the appropriate response.
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs]: >activity [](start = 45, length = 8)

Casing. See tag.

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+        /// </summary>
+        /// <param name="turnContext">Provides the <see cref="ITurnContext"/> for the turn of the bot.</param>
+        /// <returns>A <see cref="Activity"/> to send as a response to the user's input.</returns>
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs]: > <see cref="Activity"/ [](start = 22, length = 22)

See, ya have it perfect here! :)

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+            }
+            else if (activity.Text.StartsWith("2"))
+            {
+                reply.Text = "This is an attachment from a https url";
+                reply.Attachments = new List<Attachment>() { GetInternetAttachment() };
+            }
+            else if (activity.Text.StartsWith("3"))
+            {
+                reply.Text = "This is an uploaded attachment";
+                reply.Attachments = new List<Attachment>() { GetUploadedAttachmentAsync(reply.ServiceUrl, reply.Conversation.Id).Result };
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs]: >GetUploadedAttachmentAsync [](start = 61, length = 26)

Weird to see an Async method here, but no await. Please confirm this is correct, and either fix or add a comment.

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+            }
+            else
+            {
+                // The user did not enter input that this bot was built to handle.
+                reply.Text = "Your input was not recognized please try again";
+            }
+        }
+
+        /// <summary>
+        /// Handles when the user uploads an attachment to the bot by saving the attachment to
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs]: >Handles [](start = 12, length = 7)

General language cleanup needed.

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+        /// the directory this file resides in.
+        /// </summary>
+        /// <remarks>Not all channel allow users to upload files.  Some channels may have restrictions
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs]: >channel [](start = 29, length = 7)

plural. channels

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+        /// on file type, size, and other attributes.  Consult the documentation for the channel for
+        /// more information.</remarks>
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs]: >remarks [](start = 31, length = 7)

Great comment. Add the link to the docs here.

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs [hunk 18]
```diff
@@ -0,0 +1,272 @@
...
+        }
+
+        /// <summary>
+        /// Creates an inline attachment sent from the bot to the user using a base64 string.
+        /// </summary>
+        /// <returns>An <see cref="Attachment"/> to be displayed to the user.</returns>
+        /// <remarks>Using a base64 string to send an attachment will not work on all channels.
+        /// Additionally, some channels will only allow certain file types to be sent this way.
+        /// For example a .png file may work but a .pdf file may not on some channels.
+        /// Please consult the channel documentation for specifics.</remarks>
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs]: >remarks [](start = 69, length = 7)

Awesome comment. Add doc link.

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs [hunk 19]
```diff
@@ -0,0 +1,272 @@
...
+            return new Attachment
+            {
+                Name = "architecture-resize.png",
+                ContentType = "image/png",
+                ContentUrl = $"data:image/png;base64,{imageData}",
+            };
+        }
+
+        /// <summary>
+        /// Creates an attachmment to be sent from the bot to the user from an uploaded file.
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs]: >attachmment [](start = 23, length = 11)

I'm on the fence. Does this need a see tag?

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs [hunk 20]
```diff
@@ -0,0 +1,272 @@
...
+                {
+                    Name = "architecture-resize.png",
+                    ContentType = "image/png",
+                    ContentUrl = attachmentUri,
+                };
+            }
+        }
+
+        /// <summary>
+        /// Creates an attachmment to be sent from the bot to the user from a https url.
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/AtachmentsBot.cs]: >attachmment [](start = 23, length = 11)

spelling. attachment

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/Handling_Attachments.csproj [hunk 21]
```diff
@@ -0,0 +1,16 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.0</TargetFramework>
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/Handling_Attachments.csproj]: You might need the CodeAnalysis tags in here to get the Analyzers to run. Check with @daveta to get them.

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/readme.md [hunk 22]
```diff
@@ -0,0 +1,30 @@
+﻿This sample demonstrates different ways a bot may send and receive attachments.
+# Concepts introduced in this sample
+## What is a bot?
+A bot is an app that users interact with in a conversational way using text, graphics (cards), or speech. It may be a simple question and answer dialog,
+or a sophisticated bot that allows people to interact with services in an intelligent manner using pattern matching,
+state tracking and artificial intelligence techniques well-integrated with existing business services.
+## Attachments
+A message exchange between user and bot may contain cards and media attachments, such as images, video, audio, and files.
+The types of attachments that may be sent and recieved varies by channel. Additionally, a bot may also recieve file attachments.
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/readme.md]: >varies by chan [](start = 55, length = 14)

Great. Can you include a docs link here?

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/readme.md [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+## To try this sample
+- Clone the repository.
+```bash
+git clone https://github.com/microsoft/botbuilder-samples.git
+```
+ ### Visual studio
+- Navigate to the samples folder (`BotBuilder-Samples\csharp_dotnetcore\16.Handling-Atachments\Handling_Attachments`) and open Handling_Attachments.csproj in Visual Studio
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/readme.md]: >Handling_Attachments [](start = 95, length = 20)

Is this directory needed? Seems it wouldn't be.

Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/readme.md]: >Atachments [](start = 84, length = 10)

spelling. "Attachments"

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/readme.md [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+The types of attachments that may be sent and recieved varies by channel. Additionally, a bot may also recieve file attachments.
+## To try this sample
+- Clone the repository.
+```bash
+git clone https://github.com/microsoft/botbuilder-samples.git
+```
+ ### Visual studio
+- Navigate to the samples folder (`BotBuilder-Samples\csharp_dotnetcore\16.Handling-Atachments\Handling_Attachments`) and open Handling_Attachments.csproj in Visual Studio 
+- Hit F5
+ ### Visual studio code
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/readme.md]: >tudio code [](start = 13, length = 10)

Casing. Visual Studio Code.

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/readme.md [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+- Open `BotBuilder-Samples\csharp_dotnetcore\16.Handling-Atachments\Handling_Attachments` folder
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/readme.md]: >Atachments [](start = 57, length = 10)

Spelling. Attachments

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/readme.md [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+- Bring up a terminal, navigate to BotBuilder-Samples\csharp_dotnetcore\16.Handling-Atachments\Handling_Attachments
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/readme.md]: >Atachments [](start = 84, length = 10)

spelling

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/readme.md [hunk 27]
```diff
@@ -0,0 +1,30 @@
...
+- Type 'dotnet run'.
+## Update packages
+- In Visual Studio right click on the solution and select "Restore NuGet Packages".
+  **Note:** this sample requires `Microsoft.Bot.Builder`, `Microsoft.Bot.Builder.Dialogs`, `Microsoft.Bot.Connector`, and `Microsoft.Bot.Builder.Integration.AspNet.Core`.
+# Further reading
+- [Azure Bot Service Introduction](https://docs.microsoft.com/en-us/azure/bot-service/bot-service-overview-introduction?view=azure-bot-service-4.0)
+- [Bot basics](https://docs.microsoft.com/en-us/azure/bot-service/bot-builder-basics?view=azure-bot-service-4.0)
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/readme.md]: >basics [](start = 7, length = 6)

Casing on all of these is weird.

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/Startup.cs [hunk 28]
```diff
@@ -0,0 +1,97 @@
...
+            services.AddBot<AtachmentsBot>(options =>
+            {
+                options.CredentialProvider = new ConfigurationCredentialProvider(this.Configuration);
+
+                // Memory Storage is for local bot debugging only. When the bot
+                // is restarted, anything stored in memory will be gone.
+                IStorage dataStore = new MemoryStorage();
+
+                // For production bots use the Azure Blob or
+                // Azure CosmosDB storage providers, as seen below. To the Azure
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/Startup.cs]: >To th [](start = 68, length = 5)

grammar. "To use the".

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/Startup.cs [hunk 29]
```diff
@@ -0,0 +1,97 @@
...
+                // based storage providers, add the Microsoft.Bot.Builder.Azure
+                // Nuget package to your solution. That package is found at:
+                // https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/
+                // Uncomment this line to use Azure Blob Storage
+                // IStorage dataStore = new Microsoft.Bot.Builder.Azure.AzureBlobStorage("AzureBlobConnectionString", "containerName");
+                // Create and add conversation state.
+                var convoState = new ConversationState(dataStore);
+                options.State.Add(convoState);
+
+                // The BotStateSet middleware forces state storage to auto-save when the bot is complete processing the message.
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/Startup.cs]: > is compl [](start = 92, length = 9)

grammar. "has completed" or "completes"

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/Startup.cs [hunk 30]
```diff
@@ -0,0 +1,97 @@
...
+                // Note: Developers may choose not to add all the state providers to this middleware if save is not required.
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/Startup.cs]: >all the [](start = 58, length = 7)

remove "all the"

Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/Startup.cs]: >if sav [](start = 101, length = 6)

if "auto-save"

File: csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/Startup.cs [hunk 31]
```diff
@@ -0,0 +1,97 @@
...
+                var stateSet = new BotStateSet(options.State.ToArray());
+                options.Middleware.Add(stateSet);
+            });
+        }
+
+        /// <summary>
+        /// This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
+        /// </summary>
+        /// <param name="app">The <see cref="IApplicationBuilder"/>.  This provides the mechanisms to configure an application's request pipeline.</param>
```
Comment (cleemullins) [inline csharp_dotnetcore/16.Handling-Atachments/Handling_Attachments/Startup.cs]: >   [](start = 68, length = 2)

too many spaces.

**Comment by JasonSowers** [comment]
@cleemullins 👍

**Pull request merged**